pkg/machine/builder.go

```diff
@@ -0,0 +1,244 @@
+	if err != nil {
+		return nil, fmt.Errorf("failed to apply cluster information to the network config: %w", err)
+	}
+
+	return CreateProviderConfig(cloudProvider, cloudProviderSpec, operatingSystemSpec, networkConfig, nil)
+}
+
+func (b *MachineBuilder) BuildProviderSpec() (*clusterv1alpha1.ProviderSpec, error) {
+	providerConfig, err := b.BuildProviderConfig()
+	if err != nil {
+		return nil, err
+	}
+
+	return CreateProviderSpec(providerConfig)
+}
+
+func (b *MachineBuilder) determineCloudProvider() (kubermaticv1.ProviderType, error) {
+	var provider kubermaticv1.ProviderType
+
+	if b.cluster != nil {
+		clusterProvider, err := kubermaticv1helper.ClusterCloudProviderName(b.cluster.Spec.Cloud)
+		if err != nil {
+			return "", fmt.Errorf("failed to determine cloud provider from cluster: %w", err)
+		}
+
+		provider = kubermaticv1.ProviderType(clusterProvider)
+	}
+
+	// was an explicit provider given (because the caller did not have a Cluster object at hand)?
+	if b.cloudProvider != "" {
+		if provider != "" && b.cloudProvider != provider {
+			return "", fmt.Errorf("explicit cloud provider %q does not match cluster's cloud provider %q", b.cloudProvider, provider)
+		}
+
+		provider = b.cloudProvider
+	}
+
+	// all we have is a spec?
+	if b.cloudProviderSpec != nil {
+		specProvider, err := ProviderTypeFromSpec(b.cloudProviderSpec)
+		if err != nil {
+			return "", fmt.Errorf("cannot determine cloud provider for given spec: %w", err)
+		}
+
+		if provider != "" && specProvider != provider {
+			return "", fmt.Errorf("cloud provider %q from spec does not match cluster's cloud provider %q", specProvider, provider)
+		}
+
+		provider = specProvider
+	}
+
+	return provider, nil
+}
+
+func (b *MachineBuilder) determineDatacenter() (*kubermaticv1.Datacenter, error) {
+	if b.datacenter != nil {
+		return b.datacenter, nil
+	}
+
+	// sanity check: do not silently ignore misconfigurations where only a seed, but
+	// no datacenter is specified accidentally; note that the datacenter name can be
+	// taken from the cluster, if given
+	seedSpecified := b.seed != nil
+	dcSpecified := b.datacenterName != "" || (b.cluster != nil && b.cluster.Spec.Cloud.DatacenterName != "")
+
+	if seedSpecified != dcSpecified {
+		return nil, errors.New("invalid builder configuration, must specify either both seed and datacenter name or nothing at all; datacenter name can be deduced from a Cluster object")
+	}
+
+	// it's valid to not specify if the cloud provider spec needs no further fields added to it
+	if !seedSpecified && !dcSpecified {
+		return nil, nil
+	}
+
+	datacenterName := b.datacenterName
+	if datacenterName == "" && b.cluster != nil {
+		if b.cluster.Spec.Cloud.DatacenterName != b.datacenterName {
+			return nil, fmt.Errorf("explicit datacenter %q does not match cluster's datacenter %q", b.datacenterName, b.cluster.Spec.Cloud.DatacenterName)
+		}
+		datacenterName = b.cluster.Spec.Cloud.DatacenterName
+	}
```

Comment:
I'm not entirely sure if I understand this part of code. I would expect a different condition in the external `if` statement. Please double check me on that.
```suggestion
	datacenterName := b.datacenterName
	if datacenterName != "" && b.cluster != nil {
		if b.cluster.Spec.Cloud.DatacenterName != b.datacenterName {
			return nil, fmt.Errorf("explicit datacenter %q does not match cluster's datacenter %q", b.datacenterName, b.cluster.Spec.Cloud.DatacenterName)
		}
		datacenterName = b.cluster.Spec.Cloud.DatacenterName
	}
```

Reply:
Damn you're good. Thaat was indeed broken. I added some unit tests and fixed the function, should now be a bit easier to read.